Improve Command Line and add validation. v3.0

codeBelt opened the pull request:
#41
#31

codeBelt committed:
generateTemplateFilesCommandLine and output.overwrite

codeBelt committed:
generateTemplateFilesCommandLine

codeBelt committed:
clean up

codeBelt committed:
update packages

codeBelt committed:
Update README

krambono commented:
I found some "problems" that I had fixed in my previous closed pr. 
- If you don't use the overwrite flag, even if the files don't exist, it will throw an error
- If you define many times the same string replacer in command line, it doesn't work
- Errors that are throwed are not catch and sometimes the program break with the "UnhandledPromiseRejectionWarning" which is a deprecated behavior

codeBelt committed:
fix issue with files not being over written

codeBelt committed:
update packages

codeBelt committed:
errorMessageAndExit

codeBelt commented:
@knona Thanks for the feedback.

- Fixed the issue with file not writing even if they don't exist.
- Addressed the `UnhandledPromiseRejectionWarning` issue and cleaned up how errors are displayed.

I don't understand the following. It should display an error if the command line has the same string replacers in command line and not work.

> * If you define many times the same string replacer in command line, it doesn't work

The user should see this error:

<img width="985" alt="Screenshot 2020-05-23 15 26 56" src="https://user-images.githubusercontent.com/2893563/82739967-fb069300-9d09-11ea-9ba6-f1f99b3645e8.png">

codeBelt committed:
prettier

codeBelt committed:
prettier

krambono commented:
We talked about that with @ccheney. 
We could imagine that we take the last value defined of the string replacer. 
To me, both behavior are correct. 

I readed your last commit. I saw you use the process.exit() function. Why don't you catch the errors with a try / catch ?

From what I know using the exit function is not recommended.

codeBelt committed:
fix and skip tests

codeBelt committed:
Merge branch 'command-line-3.0' into prettier

# Conflicts:
#	src/utilities/CheckUtility.test.ts
#	src/utilities/CheckUtility.ts
#	src/utilities/GenerateTemplateFiles.commandline.test.ts

codeBelt commented:
Ok, I changed the `process.exit()` to a `try/catch`. Thanks @knona for reviewing this.

codeBelt commented on src/utilities/CheckUtility.ts:
```diff
@@ -4,51 +4,34 @@ import StringUtility from './StringUtility';
+            throw new Error(errorMessage);
+        } catch (error) {
+            console.info(colors.bold.red(`[Error in generate-template-files]: ${colors.red(error.message)}`));
+        }
```
@knona @ccheney I didn't want to put a `try/catch` everywhere so I did this. Seems like hacky but it works.

krambono replied:
If you throw the error in the try block, you will always jump to the catch block. Therefore, the code is equivalent to the code in the catch block, without throwing an error. 
You should use the try...catch at the top of the call stack.
Otherwise the code won't break and will continue to run normally.
Ex: do a ctrl + c during an equirer prompt. 

You should set the try catch in the main methods which are "generate" and "commandLine". 
To me it's the best way to handle the errors even if it is a bit ugly.

codeBelt replied:
done

codeBelt commented:
> We could imagine that we take the last value defined of the string replacer.

@knona to make things easier we should just error and have the user fix their mistake then trying to fix it for them.

krambono commented:
> > We could imagine that we take the last value defined of the string replacer.
> 
> @knona to make things easier we should just error and have the user fix their mistake then trying to fix it for them.
> 
> 

I agree.

codeBelt committed:
Change to throw errors and add try/catch. add display* functions

codeBelt committed:
Merge branch 'command-line-3.0' into prettier

# Conflicts:
#	src/GenerateTemplateFiles.ts
#	src/utilities/CheckUtility.test.ts
#	src/utilities/CheckUtility.ts

codeBelt committed:
Merge pull request #45 from codeBelt/prettier

Prettier
